### PR TITLE
deduplicate coin states

### PR DESCRIPTION
**Pull request by ytx1991**
### Purpose:
In some situations, one coin can be processed multiple times. To prevent that, we need to add deduplication in add_coin_states.

### Current Behavior:
No deduplication

### New Behavior:
Deduplicate  coin states

### CODE REVIEW

File: chia/wallet/wallet_state_manager.py [hunk 1]
```diff
@@ -1602,7 +1603,8 @@ async def add_coin_states(
         fork_height: Optional[uint32],
     ) -> bool:
         try:
-            await self._add_coin_states(list(set(coin_states)), peer, fork_height)
+            unique_coin_states = sort_coin_states(list(set(coin_states)))
```
Comment (arvidn) [inline chia/wallet/wallet_state_manager.py:1606]: both before and after does this:
`list(set(coin_states))`
which is deduplicating coins. Can you explain how this change makes a difference in deduplication?

**Review by arvidn — COMMENTED** [review]
it looks like you just order the coin states, they were already deduplicated

File: chia/wallet/wallet_state_manager.py (hunk 1 above)
Comment (ytx1991) [reply chia/wallet/wallet_state_manager.py:1606]: I can't find the deduplicating code you are referring to, can you add a link?

Comment (arvidn) [reply chia/wallet/wallet_state_manager.py:1606]: putting items in a `set` deduplicates them. We do that on the line I put this comment on.

File: tests/wallet/clawback/test_clawback_lifecycle.py [hunk 2]
```diff
@@ -195,9 +195,10 @@ async def test_clawback_spends(self, cost_logger: CostLogger) -> None:
                 cost_logger=cost_logger,
                 cost_log_msg="Create Second Clawback",
             )
-
-            new_cb_coin = (await sim_client.get_coin_records_by_puzzle_hash(cb_puz_hash))[1].coin
-
+            cb_coins = await sim_client.get_coin_records_by_puzzle_hash(cb_puz_hash)
+            new_cb_coin = cb_coins[0].coin
+            if cb_coins[0].spent:
+                new_cb_coin = cb_coins[1].coin
```
Comment (arvidn) [inline tests/wallet/clawback/test_clawback_lifecycle.py:201]: please add a comment to what this logic intends to do, and maybe why it's safe to assume that the second coin in this list is the one you want

File: chia/wallet/wallet_state_manager.py (hunk 1 above)
Comment (ytx1991) [reply chia/wallet/wallet_state_manager.py:1606]: Oh I see, you need to check the diff with the release/1.8.2. It doesn't have List(set()) thing there.

Comment (xdustinface) [reply chia/wallet/wallet_state_manager.py:1606]: > putting items in a `set` deduplicates them. We do that on the line I put this comment on.

You only look at the second commit, i guess?

File: chia/wallet/wallet_state_manager.py [hunk 3]
```diff
@@ -1602,7 +1603,8 @@ async def add_coin_states(
         fork_height: Optional[uint32],
     ) -> bool:
         try:
-            await self._add_coin_states(coin_states, peer, fork_height)
+            unique_coin_states = sort_coin_states(list(set(coin_states)))
+            await self._add_coin_states(unique_coin_states, peer, fork_height)
```
Comment (xdustinface) [inline chia/wallet/wallet_state_manager.py:1607]: This isn't the right place to deduplicate the coin states imo because this is after we already sorted all coin states here https://github.com/Chia-Network/chia-blockchain/blob/b8a1515d7350f92e246a369b2685fafdc7864d5c/chia/wallet/wallet_node.py#L870 before we split them up into batches to process in tasks which call `add_coin_states` here https://github.com/Chia-Network/chia-blockchain/blob/b8a1515d7350f92e246a369b2685fafdc7864d5c/chia/wallet/wallet_node.py#L905 

So basically you now only deduplicate the separate batches instead the whole list coming from a peer and we would still validate duplicates and also this adds a second round of sorting.

**Review by xdustinface — CHANGES_REQUESTED** [review]

File: chia/wallet/wallet_state_manager.py (hunk 3 above)
Comment (xdustinface) [reply chia/wallet/wallet_state_manager.py:1607]: I just created https://github.com/Chia-Network/chia-blockchain/pull/15598.

File: tests/wallet/clawback/test_clawback_lifecycle.py (hunk 2 above)
Comment (xdustinface) [reply tests/wallet/clawback/test_clawback_lifecycle.py:201]: I also wonder, whats the purpose of this change?

File: chia/wallet/wallet_state_manager.py (hunk 1 above)
Comment (arvidn) [reply chia/wallet/wallet_state_manager.py:1606]: yes. I clicked on the link in the email and apparently only saw a single commit

**Comment by ytx1991** [comment]
Close this PR since it is solved in https://github.com/Chia-Network/chia-blockchain/pull/15598
